### PR TITLE
Viewboxed combine separate

### PR DESCRIPTION
**Pull request by seemantk**
@curran, as hoped for combine/separate has been simplified!

### CODE REVIEW

**Comment by curran** [comment]
Nice!

**Comment by seemantk** [comment]
OK, fixed the zoom/pan bug after I realized the obvious: nav affects x, combine/separate affects y.  Thus, they now just leave the other axis unchanged.

**Comment by seemantk** [comment]
(Note, I don't know if I'll be able to show/hide ribbon and ribbon types on time for our call)

**Comment by seemantk** [comment]
@curran. ok I think the zoom/pan + combine/separate issue has been sorted.
the ribbon controls have also been wired up (as a side-effect, mainly, because I needed to test).

Edit: forgot to mention: I updated the controls to a more toggle-style approach.  Except for the "notes" floating button.  Haven't landed on an intuitive controller for:

- show/hide notes, but only hide if there are ribbons showing
- show/hide extremes

This is a merge-candidate, knowing that the reference lines and mensurations signs are still missing.

**Comment by curran** [comment]
@seemantk Fantastic! Will have a look.

**Comment by curran** [comment]
This looks and feels great. I'd be happy to merge this. I'd also like to keep a copy of the old UI available for the Josquin team to play with. I think what I'll do is take the current gh-pages version and give it the subdirectory treatment (within this branch probably), then merge.

File: js/baton.js [hunk 1]
```diff
@@ -1,53 +1,67 @@
-var width = 960
-  , height = 150 // height of one strip of notes
-  , margin = { top: 20, right: 20, bottom: 20, left: 20 }
+var margin = { top: 20, right: 20, bottom: 20, left: 20 }
+  , divMeta = d3.select("#meta")
+  , canvas = NotesCanvas()
+      .svg(d3.select("body").append("svg").attr("id", "unrendered"))
```
Comment (curran) [inline js/baton.js:4]: Now entering the shadow realm :+1:

File: js/baton.js [hunk 2]
```diff
@@ -1,53 +1,67 @@
...
   , notesUI = NotesUI()
       .div(divMeta.select("#notes-ui"))
-  , ribbonsUI = RibbonsUI()
-      .div(divMeta.select("#ribbons-ui"))
   , colorLegend = ColorLegend()
       .div(divMeta.select("#legend"))
   , colorScale = d3.scaleOrdinal(d3.schemeCategory10)
-  , lifeSize = 10 // screen width of 1 duration
-  , lifeScale = d3.scaleLinear()
+  , combineButton = [{
```
Comment (curran) [inline js/baton.js:12]: I love the switches here, very cool.

File: js/baton.js [hunk 3]
```diff
@@ -1,53 +1,67 @@
...
         part.notedata.forEach(function(note) {
+            var pitch = note.pitch.b7;
             notes.push({
-                  pitch: note.pitch.b7
+                  pitch: pitch
                 , note: note.pitch.name
                 , time: note.starttime[0]
                 , duration: note.duration[0]
-                , voice: voice
+                , extreme: pitch === extent[0] || pitch === extent[1]
```
Comment (curran) [inline js/baton.js:56]: Awesome to compute `extreme` here! Paves the way to easily add another field `globalExtreme` here for #44 .

File: js/baton.js [hunk 4]
```diff
@@ -59,33 +73,42 @@ function chartify(data) {
...
-        .colorScale(colorScale)
-        .margin(margin)
-        .height(height * 3)
-        .width(width)
-        .extremes(true)
-        .data(data)
-        .connect(signal)
+    d3.select("#separate-ui")
+        .datum(combineButton)
+        .call(ToggleUI().connect(signal))
```
Comment (curran) [inline js/baton.js:103]: Nice call pattern.

File: js/baton.js [hunk 5]
```diff
@@ -152,3 +161,15 @@ function getQueryVariables() {
     ;
     return inits;
 } // getQueryVariables()
+
+// Convert a string into a form suitable for use as a css class name
+// Hopefully the voice names don't contain Unicode strings
+// From: https://gist.github.com/mathewbyrne/1280286#gistcomment-1716050
+function slugify(str) {
```
Comment (curran) [inline js/baton.js:168]: Pretty neat! Candidate for its own little library.

File: js/notesbook.js [hunk 6]
```diff
@@ -217,150 +59,97 @@ function NotesBook() {
...
-    } // my.reset()
-  my.extremes = function(value) {
-      if(!arguments.length) return display.extremes;
-
-      display.extremes = value;
-      canvases.forEach(function(c) {
-          c.canvas.extremes(value);
-      });
-
+  my.separate = function(_) {
```
Comment (curran) [inline js/notesbook.js:113]: Viewboxy goodness!

File: js/notescanvas.js [hunk 7]
```diff
@@ -284,81 +93,77 @@ function NotesCanvas() {
...
+          .append("svg")
+            .call(sizeit)
+            .attr("class", "lens")
+             // Stretch contents to container
+            .attr("preserveAspectRatio", "none")
+            .style("width", "100%")
+            .style("height", "100%")
+          .merge(sheet)
+        ;
+        // Nest the voice SVGs
```
Comment (curran) [inline js/notescanvas.js:138]: Interesting approach to have nested SVGs.

File: js/ribbon.js [hunk 8]
```diff
@@ -1,126 +1,119 @@
...
-                , noteStart = d.time
-                , noteEnd   = d.time + d.duration
+        ribbon.exit().remove();
+        ribbon
+          .enter().append("g")
+            .attr("class", function(d) { return d.toLowerCase(); })
+          .merge(ribbon)
+          .each(function(m) {
+              var path = d3.select(this).selectAll("path")
+                  .data([modes[m](data)]) // call the mode function
```
Comment (curran) [inline js/ribbon.js:35]: Nice way of splitting out mode definitions.

File: js/score.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+function Score() {
```
Comment (curran) [inline js/score.js:1]: So simple, this is awesome!

File: js/score.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  my.x = function(_) {
+      if(!arguments.length) return x;
+      x = _;
+      return my;
+    } // my.x()
+  ;
+  my.y = function(_) {
+      if(!arguments.length) return y;
+      y = _;
+      noteHeight = y.bandwidth();
```
Comment (curran) [inline js/score.js:46]: Small thing, but maybe this line `noteHeight = y.bandwidth();` could be moved inside `my()` as `var noteHeight = y.bandwidth();`

**Review by curran — APPROVED** [review]
Awesome work here @seemantk .

**Pull request merged**